### PR TITLE
Add draft update proposal for update to P4.

### PR DESCRIPTION
**Pull request by abizjak**
## Purpose

Addresses #21

### CODE REVIEW

File: updates/P4.txt [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+the smart contract instance. The behaviour of `InitContract` and
+`Update` transactions is dependent on the module version. Version 1
+contracts can access the new contract state, can call other contracts
+and synchronously transfer CCD to an account. Version 1 contracts also
+have access to cryptographic primitives `verify_ed25519_signature`,
+`verify_ecdsa_secp256k1_signature`, `hash_sha2_256`, `hash_sha3_256`,
+and `hash_keccak_256`. Finally, version 1 contracts can have a
+fallback entrypoint, which is invoked if none of the other entrypoints
+in the contract matches. The semantics of version 0 contracts stays
+exactly as it was in protocol versions 1, 2, and 3. directly
```
Comment (td202) [inline updates/P4.txt:83]:
```suggestion
exactly as it was in protocol versions 1, 2, and 3.
```

**Review by td202 — CHANGES_REQUESTED** [review]
Generally fine. I've requested a few changes to be a bit more specific about some of the delegation changes.

File: updates/P4.txt [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+With regards to smart contracts, in protocol versions 1, 2, and 3
+smart contracts have limited state (16kB) and communicate via message
+passing. The former limitation prevents use-cases such as exchanges,
+or stablecoins which require a large database. The choice of
+message-passing is not a technical limitation, but it does make
+contracts harder to write. In protocol version 4 the state will not
+have an explicit upper limit, however its growth is still limited by
+the costs. The message passing communication is replaced by direct
+contract calls.
```
[the remark this answers is not in the document]
Comment (abizjak) [reply updates/P4.txt]: Added.

**Comment by abizjak** [comment]
> Generally fine. I've requested a few changes to be a bit more specific about some of the delegation changes.

I have addressed (well, mostly just accepted the suggestions) the comments. 

Unless there are further comments by tomorrow at 11 I will update the hash and include it in the node.

**Pull request merged**